src/Core/Services/AuthService.cs

```diff
@@ -198,12 +198,39 @@ private async Task<bool> RequirePasswordChangeAsync(string email, string masterP
             return !await _policyService.EvaluateMasterPassword(strength.Value, masterPassword, _masterPasswordPolicy);
         }
 
-        public async Task<AuthResult> LogInPasswordlessAsync(string email, string accessCode, string authRequestId, byte[] decryptionKey, string userKeyCiphered, string localHashedPasswordCiphered)
+        public async Task<AuthResult> LogInPasswordlessAsync(string email, string accessCode, string authRequestId, byte[] decryptionKey, string encryptedAuthRequestKey, string masterKeyHash)
         {
-            var decKey = await _cryptoService.RsaDecryptAsync(userKeyCiphered, decryptionKey);
-            var decPasswordHash = await _cryptoService.RsaDecryptAsync(localHashedPasswordCiphered, decryptionKey);
-            return await LogInHelperAsync(email, accessCode, Encoding.UTF8.GetString(decPasswordHash), null, null, null, new MasterKey(decKey), null, null,
-                null, null, authRequestId: authRequestId);
+            AuthResult response = null;
+            // On SSO flow user is already AuthN
+            if (await _stateService.IsAuthenticatedAsync())
+            {
+                var decryptedKey = await _cryptoService.RsaDecryptAsync(encryptedAuthRequestKey, decryptionKey);
+                if (string.IsNullOrEmpty(masterKeyHash))
+                {
+                    await _cryptoService.SetUserKeyAsync(new UserKey(decryptedKey));
+                }
+                else
+                {
+                    var userKey = await _cryptoService.DecryptUserKeyWithMasterKeyAsync(new MasterKey(decryptedKey));
+                    await _cryptoService.SetUserKeyAsync(userKey);
+                }
+                await _deviceTrustCryptoService.TrustDeviceIfNeededAsync();
+            }
+            else
+            {
+                var decryptedKey = await _cryptoService.RsaDecryptAsync(encryptedAuthRequestKey, decryptionKey);
+                if (string.IsNullOrEmpty(masterKeyHash) && decryptionKey != null)
+                {
+                    await _cryptoService.SetUserKeyAsync(new UserKey(decryptedKey));
+                }
+                else
+                {
+                    var decKeyHash = await _cryptoService.RsaDecryptAsync(masterKeyHash, decryptionKey);
+                    response = await LogInHelperAsync(email, accessCode, Encoding.UTF8.GetString(decKeyHash), null, null, null, new MasterKey(decryptedKey), null, null,
+                    null, null, authRequestId: authRequestId);
+                }
+            }
+            return response;
```

Comment:
⛏️  Sorry I missed this in the previous review, but no need to have the `response` object be created before, also the `decryptedKey`  can be moved outside the if (now it's being repeated) and finally, we can do the next thing to improve arrow code.
```suggestion
            var decryptedKey = await _cryptoService.RsaDecryptAsync(encryptedAuthRequestKey, decryptionKey);
            
            // On SSO flow user is already AuthN
            if (await _stateService.IsAuthenticatedAsync())
            {
                if (string.IsNullOrEmpty(masterKeyHash))
                {
                    await _cryptoService.SetUserKeyAsync(new UserKey(decryptedKey));
                }
                else
                {
                    var userKey = await _cryptoService.DecryptUserKeyWithMasterKeyAsync(new MasterKey(decryptedKey));
                    await _cryptoService.SetUserKeyAsync(userKey);
                }
                await _deviceTrustCryptoService.TrustDeviceIfNeededAsync();
                return null;
            }

            if (string.IsNullOrEmpty(masterKeyHash) && decryptionKey != null)
            {
                await _cryptoService.SetUserKeyAsync(new UserKey(decryptedKey));
                return null;
            }
            
            var decKeyHash = await _cryptoService.RsaDecryptAsync(masterKeyHash, decryptionKey);
            return await LogInHelperAsync(email, accessCode, Encoding.UTF8.GetString(decKeyHash), null, null, null, new MasterKey(decryptedKey), null, null,
                null, null, authRequestId: authRequestId);
```

Reply:
Beautiful!